indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -575,58 +575,234 @@ public Response getCompleteTasks(
+        maxCompletedTasks, duration);
+    for (Task task : tasks) {
+      completeTaskCache.putIfAbsent(task.getId(), task);
+    }
 
-    final List<TaskStatus> recentlyFinishedTasks = Lists.newArrayList(
-        AuthorizationUtils.filterAuthorizedResources(
-            req,
-            taskStorageQueryAdapter.getRecentlyFinishedTaskStatuses(maxTaskStatuses),
-            raGenerator,
-            authorizerMapper
-        )
-    );
+    final List<TaskInfo> taskInfoList = taskStorageQueryAdapter.getRecentlyCompletedTaskInfo(
+        maxCompletedTasks, duration);
+    for (TaskInfo taskInfo : taskInfoList) {
+      taskInfoCache.putIfAbsent(taskInfo.getId(), taskInfo);
+    }
+
+    final List<TaskStatus> recentlyFinishedTasks = taskStorageQueryAdapter.getRecentlyFinishedTaskStatuses(
+        maxCompletedTasks, duration);
 
     final List<TaskStatusPlus> completeTasks = Lists.newArrayList(Iterables.transform(
         recentlyFinishedTasks,
         status -> {
-          final Pair<DateTime, String> pair = taskStorageQueryAdapter.getCreatedDateAndDataSource(status.getId());
           return new TaskStatusPlus(
               status.getId(),
-              taskFunction.apply(status.getId()).getType(),
-              pair.lhs,
+              completeTaskCache.get(status.getId()).getType(),
+              taskInfoCache.get(status.getId()).getCreatedTime(),
               // Would be nice to include the real queue insertion time, but the
               // TaskStorage API doesn't yet allow it.
               DateTimes.EPOCH,
               status.getStatusCode(),
+              RunnerTaskState.NONE,
               status.getDuration(),
               TaskLocation.unknown(),
-              pair.rhs,
+              taskInfoCache.get(status.getId()).getDataSource(),
               status.getErrorMsg()
           );
-        }));
+        }
+    ));
+
+    return completeTasks;
+  }
+
+  private List<AnyTask> filterActiveTasks(HttpServletRequest req, RunnerTaskState state, List<Task> allActiveTasks)
+  {
+
+    final List<AnyTask> allTasks = Lists.newArrayList();
+    for (final Task task : allActiveTasks) {
+      allTasks.add(
+          new AnyTask(
+              task.getId(),
+              task.getType(),
+              SettableFuture.create(),
+              task.getDataSource(),
+              null,
+              null,
+              DateTimes.EPOCH,
+              DateTimes.EPOCH
+          ));
+    }
+
+    //divide active task into 3 lists : running, pending, waiting
+    Optional<TaskRunner> taskRunnerOpt = taskMaster.getTaskRunner();
+    if (!taskRunnerOpt.isPresent()) {
+      throw new WebApplicationException(
+          Response.serverError().entity("No task runner found ").build()
+      );
+    }
+    TaskRunner runner = taskRunnerOpt.get();
+    Collection<? extends TaskRunnerWorkItem> runnersKnownTasks = runner.getKnownTasks();
+    List<String> runnerKnownTaskIds = runnersKnownTasks.stream()
```

Comment:
Since you're calling `contains` on this a lot, it should be a Set. `contains` on Lists is generally O(n), whereas `contains` on Sets is generally O(log n) or O(1).

Reply:
done